content/blog/how-to-use-nodejs-loader.en.mdx

```diff
@@ -0,0 +1,289 @@
+
+By defining these hooks, the loader can customize how Node.js handles different types of modules, making it a powerful tool for developers. This customization can be particularly useful for tasks such as:
+
+- Loading non-JavaScript files (e.g., TypeScript, CoffeeScript)
+- Transpiling code on-the-fly
+- Implementing custom module formats
+- Overriding or extending the default module resolution logic
+
+### How to use Node.js ESM loader
+
+Using the Node.js ESM loader involves creating custom hooks that intercept and modify the default behavior of module resolution and loading. These hooks allow you to handle non-JavaScript files, transpile code on-the-fly, and implement custom module formats. Here's a step-by-step guide on how to use the Node.js ESM loader:
+
+#### Step 1: Create a Loader Module
+
+First, you need to create a module that exports the custom hooks. This module will contain the `resolve` and `load` hooks that define how modules are resolved and loaded.
+
+Here's an example of a loader module that handles `.coffee` files by transpiling them to JavaScript:
+
+```mjs
+// coffee-loader.mjs
+import { readFile } from 'node:fs/promises';
+import { dirname, extname, resolve as resolvePath } from 'node:path';
+import { cwd } from 'node:process';
+import { fileURLToPath, pathToFileURL } from 'node:url';
+import coffeescript from 'coffeescript';
+
+const extensionsRegex = /\.(coffee|litcoffee|coffee\.md)$/;
+
+export async function load(url, context, nextLoad) {
+  if (extensionsRegex.test(url)) {
+    const format = await getPackageType(url);
+    const { source: rawSource } = await nextLoad(url, { ...context, format });
+    const transformedSource = coffeescript.compile(rawSource.toString(), url);
+
+    return {
+      format,
+      shortCircuit: true,
+      source: transformedSource,
+    };
+  }
+
+  return nextLoad(url);
+}
+
+async function getPackageType(url) {
+  const isFilePath = !!extname(url);
+  const dir = isFilePath ? dirname(fileURLToPath(url)) : url;
+  const packagePath = resolvePath(dir, 'package.json');
+  const type = await readFile(packagePath, { encoding: 'utf8' })
+    .then((filestring) => JSON.parse(filestring).type)
+    .catch((err) => {
+      if (err?.code !== 'ENOENT') console.error(err);
+    });
+
+  return dir.length > 1 && getPackageType(resolvePath(dir, '..'));
+}
+```
+
+#### Step 2: Register the Loader
+
+Next, you need to register the loader module using the `register` method from `node:module`. This can be done using the `--import` flag when running your Node.js application.
+
+Here's an example of how to register the loader module:
+
+```bash
+node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register(pathToFileURL("./coffee-loader.mjs"));' ./main.coffee
+```
```

Comment:
This looks rather cumbersome and nobody would do this IRL, so I would make an example that reflects what someone would do (like `--import ./register/coffee.mjs`)